daffodil-runtime1/src/main/scala/org/apache/daffodil/infoset/InfosetInputter.scala

```diff
@@ -420,7 +420,14 @@ abstract class InfosetInputter
         }
       } else if (erd.outputValueCalcExpr.isEmpty) {
         val primType = elem.erd.optPrimType.get
-        val obj = primType.fromXMLString(txt)
+        val obj = try {
+          primType.fromXMLString(txt)
```

Comment:
Grepping for ``fromXMLString``, there's a handful of other places where we call this without wrapping it with try/catch. I'm wondering if we have this same issue in other places and need the try/catch?

I'm also concerned that different primitive types use different logic in their fromXMLString implementation that might throw different exceptions that we might not be accounting for here. Thoughts on modifying the ``fromXMLString`` implementations to be more thorough about checking for exceptions, and then they rethrow a single exception that callers of ``fromXMLString`` should catch so they don't have to know about particulars of different prim type implementations? This is similar to [DAFFODIL-1681](https://issues.apache.org/jira/browse/DAFFODIL-1681).

Reply:
I did a look through all the different prim types and it looked like we would only see NumberFormatExceptions for all the various number primitives (int/short/float/double, etc) and IllegalArgurmentExceptions for HexBinary and Date/Time/Calendar types. Although I just double checked and realized I missed the URI type, which can potentially throw a NullPointerException or URISyntaxException. You are probably right that the right move is to modify all the primitive types to handle their own exceptions and then throw a more generic exception to be caught elsewhere.

Reply:
Yeah, we should make fromXMLString a tighter API which throws a specific exception when unable to convert. This is a case where Scala not enforcing handling of exceptions is not doing us a favor however. 

A perhaps more Scala-ish way of doing this is to make it return a Try[String] object instead of throwing, though we tend to avoid objects like Try, Option/Some, or little crud tuples, etc. in the runtime. Unfortunately we don't have our own non-allocating version of Try[T] as yet.

Reply:
Consider this:
`/**
 * Scala doesn't enforce that throws be handled. Hence APIs that involve throwing
 * an exception that callers are expected to always handle are fragile because callsers
 * might not handle them, and compilation doesn't tell you that.
 *
 * This solves that problem by putting the exception handling into the type signature of
 * the return.
 *
 * Note: ResultType cannot be a class derived from Exception, in case that wasn't obvious.
 */
class TryVal[ResultType <: AnyRef, ExceptionType <: Exception] (private val thing: AnyRef) extends AnyVal {
  @inline def isValue: Boolean = this.thing.isInstanceOf[ResultType]
  @inline private def value: ResultType = this.asInstanceOf[ResultType]
  @inline def isException : Boolean = this.thing.isInstanceOf[ExceptionType]
  @inline def toss = throw this.asInstanceOf[ExceptionType]
  @inline def get: ResultType = if (isValue) value else toss
}`

Reply:
Nah. Nevermind. Type tags would be needed due to type erasure, etc. etc. Too complicated.

Reply:
I've added some extra try/catch blocks in a few places, but not all of them. Some of the instances that I did not add try/catch to did not seem like it would necessarily need them. For example, there are multiple calls in XMLUtils, but those I believe are only really called in testing code. One place I think I do need to add something is in VariableUtils.convert(), but if I remember correctly we don't really want to throw exceptions from the runtime. Would the appropriate action be to catch the new exception and then do an assert?

Reply:
If it's in the scope of a parse call, then you have to catch it and return a failed status with the cause being the caught exception. In unparser you convert the exception to a UnparseError which is then rethrown.